src/Frontend/Core.hs

```diff
@@ -247,13 +251,15 @@ redirectFormHandler getPage processor = getH :<|> postH
         v <- getForm fa (processor1 page)
         renderer page v fa
 
-    postH env = do
+    postH formData = do
         page <- getPage
         let fa = absoluteUriPath $ formAction page
+            env = getFormDataEnv formData
         (v, mpayload) <- postForm fa (processor1 page) (\_ -> return $ return . runIdentity . env)
-        case mpayload of
+        (case mpayload of
             Just payload -> processor2 page payload >>= redirect
-            Nothing      -> renderer page v fa
+            Nothing      -> renderer page v fa)
+            `finally` cleanupTempCsvFiles formData
```

Comment:
Interesting!  I will add the second comment to the haddocks.

As for the strictness, what about another constraint?:

```haskell
redirectFormHandler
    :: (FormPageView p, Page p, ActionM m, NFData (FormPageResult p))
    => [...]
```

Then we can call `deepseq` on `payload`; probably after the processor has been called.  Not in the spirit of squeezing the best possible performance out of the system, but much safer.